app/services/canvas_service.py

```diff
@@ -339,61 +335,70 @@ async def create_course_folder(
             locked
         )
     
-    async def upload_grade(
+    async def upload_submission(
         self,
         assignment_id: int,
         user_id: int,
-        # not, literally, but \in [0,1]
-        grade_percent: float,
         student_notebook: BinaryIO,
-        comments: str | None = None,
+        comments: str | None = None
     ):
         url = f"courses/{ settings.CANVAS_COURSE_ID }/assignments/{ assignment_id }/submissions"
         iso_now = get_now_with_tzinfo().isoformat()
 
+        assignment = await self.get_assignment(assignment_id)
         student_course_submissions_folder = await self.get_student_course_submissions_folder_path()
-        # You can't attach course files to submissions, this is so that
-        # instructors have an alternative place to view/download submissions
         await self.upload_course_file(
             student_notebook,
-            os.path.join(student_course_submissions_folder, str(assignment_id)),
+            os.path.join(student_course_submissions_folder, assignment["name"]),
             on_duplicate=DuplicateFileAction.OVERWRITE
         )
         student_notebook_file_id = await self.upload_submission_file(
             assignment_id,
             user_id,
             student_notebook,
-            "",
             on_duplicate=DuplicateFileAction.OVERWRITE
         )
-
-        posted_grade = f"{ grade_percent * 100 }%"
         
         payload = {
             "submission": {
                 "user_id": user_id,
                 "submission_type": "online_upload",
-                "posted_grade": posted_grade,
                 "file_ids": [student_notebook_file_id],
                 "submitted_at": iso_now
             },
-            "comment": {},
-            "prefer_points_over_scheme": True
+            "comment": {}
         }
         if comments is not None: payload["comment"]["text_comment"] = comments
 
-        # Although posted_grade is listed as a supported argument of the endpoint,
-        # it only works to set the grade for the first submission by the student.
-        submission = await self._post(url, json=payload)
-        await self._put(f"{ url }/{ user_id }", json={
+        await self._post(url, json=payload)
+
+    """
+    Note that Canvas does not support associating grades with submissions.
+    You can only set the overall assignment grade for the student.
+    Comments will be attached to the active submission.
+    """
+    async def upload_assignment_grade(
+        self,
+        assignment_id: int,
+        user_id: int,
+        # between [0,1]
+        grade_proportion: float,
```

Comment:
The comment above this says it can be any value between 0 and 1. Why not enforce that with a type? Maybe like this?
```python
class GradeProportion(float):
        def __new__(cls, input = 0, *args, **kwargs):
            instance = int.__new__(cls, input, *args, **kwargs)
            if not 0 <= instance <= 1:
                raise ValueError('Value outside GradeProportion range')
            return instance
```

Reply:
Then you can only accept a `GradeProportion` here instead of any float, and you won't have to add validation within this method.

Reply:
Oh yes there's a way to do that with pydantic I think

Reply:
It's confloat with ge=0, le=1

Reply:
Looks like its use is discouraged in the docs, since it doesn't play well with static analysis tools: https://docs.pydantic.dev/latest/api/types/#pydantic.types.confloat

They give another solution, to use this
```python
class Foo(BaseModel):
    bar: Annotated[float, Field(strict=True, gt=0)]
```
which I thiiink works just as well as the class I gave above

Reply:
An annotated float works just as well.

Reply:
Sounds good. So are you going to change this float to an annotated float with constraints, or leave it as is?